server/src/main/java/com/microsoft/java/bs/core/internal/gradle/actions/GetSourceSetsAction.java

```diff
@@ -0,0 +1,111 @@
+import org.gradle.tooling.BuildController;
+import org.gradle.tooling.model.gradle.GradleBuild;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * {@link BuildAction} that retrieves {@link DefaultGradleSourceSet} from a Gradle build,
+ * handling both normal and composite builds.
+ */
+public class GetSourceSetsAction implements BuildAction<GradleSourceSets> {
+
+  /**
+   * Executes the build action and retrieves source sets from the Gradle build.
+   *
+   * @return A {@link DefaultGradleSourceSets} object containing all retrieved source sets.
+   */
+  @Override
+  public GradleSourceSets execute(BuildController buildController) {
+    var traversedProjects = new HashSet<String>();
+    var sourceSetToClasspath = new HashMap<GradleSourceSet, List<File>>();
+    var outputsToSourceSet = new HashMap<File, GradleSourceSet>();
+
+    GradleBuild buildModel = buildController.getBuildModel();
+    String rootProjectName = buildModel.getRootProject().getName();
+    fetchModels(buildController,
+        buildModel,
+        traversedProjects,
+        sourceSetToClasspath,
+        outputsToSourceSet,
+        rootProjectName);
+
+    // Add dependencies
+    var sourceSets = new ArrayList<GradleSourceSet>();
+    for (var entry : sourceSetToClasspath.entrySet()) {
+
+      var dependencies = new HashSet<BuildTargetDependency>();
+      for (File file : entry.getValue()) {
+        GradleSourceSet otherSourceSet = outputsToSourceSet.get(file);
+        if (otherSourceSet != null && !Objects.equals(entry.getKey(), otherSourceSet)) {
+          dependencies.add(new DefaultBuildTargetDependency(otherSourceSet));
+        }
+      }
+
+      DefaultGradleSourceSet sourceSet = new DefaultGradleSourceSet(entry.getKey());
+      sourceSet.setBuildTargetDependencies(dependencies);
+      sourceSets.add(sourceSet);
+
+    }
+
+    return new DefaultGradleSourceSets(sourceSets);
+  }
+
+  /**
+   * Fetches source sets from the provided Gradle build model and
+   * stores them in a map categorized by project name.
+   *
+   * @param buildController      The Gradle build controller used to interact with the build.
+   * @param build                The Gradle build model representing the current build.
+   * @param traversedProjects    A set of traversed project names to avoid cyclic dependencies.
+   * @param sourceSetToClasspath A map that associates GradleSourceSet objects with their
+   *                             corresponding classpath files.
+   * @param outputsToSourceSet   A map that associates output files with the GradleSourceSet
+   *                             they belong to.
+   * @param buildName            The name of the root project in the build.
+   */
+  private void fetchModels(
+      BuildController buildController,
+      GradleBuild build,
+      HashSet<String> traversedProjects,
```

Comment:
nit: `HashSet<String>` -> `Set<String>`

To align with the below params, since they are Map, not HashMap